### PR TITLE
sql: fix remaining AGGREGATE functions using reduce elision optimization

### PR DESCRIPTION
**Pull request by mkysel**
Fix:
- JSONB_AGG
- JSONB_OBJECT_AGG
- LIST_AGG
- ARRAY_AGG

Also move the `on_unique` functionality to `AggregateExpr` and replace the `match _` with an explicit listing of all aggregates as per #8595.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/expr/src/relation/mod.rs [hunk 1]
```diff
@@ -1542,6 +1542,99 @@ impl AggregateExpr {
             _ => self.expr.is_literal_err(),
         }
     }
+
+    /// Extracts unique input from aggregate type
+    pub fn on_unique(&self, input_type: &RelationType) -> Result<MirScalarExpr, EvalError> {
```
Comment (mkysel) [inline src/expr/src/relation/mod.rs:1547]: the error handling is actually no longer required...

File: doc/user/content/release-notes.md [hunk 2]
```diff
@@ -127,7 +127,8 @@ List new features before bug fixes.
   [`mz_kafka_source_statistics`](/sql/system-catalog#mz_kafka_source_statistics),
   containing raw statistics from the underlying librdkafka library.
 
-- Fix a bug that caused a panic when using a query containing `STRING_AGG`
+- Fix a bug that caused a panic when using a query containing
```
Comment (sploiselle) [inline doc/user/content/release-notes.md:130]: - Should end in a period
- Adding any kind of detail about when this occurred would be nominally more confidence building in release note readers.

**Review by sploiselle — APPROVED** [review]
LGTM but I'm not intimately familiar with `reduce_elision`, so +1

File: doc/user/content/release-notes.md (hunk 2 above)
Comment (mkysel) [reply doc/user/content/release-notes.md:130]: sure!

**Comment by mkysel** [comment]
adding @philip-stoev to test all aggregate functions (JSONB_AGG, JSONB_OBJECT_AGG, LIST_AGG, ARRAY_AGG, STRING_AGG)

File: doc/user/content/release-notes.md [hunk 3]
```diff
@@ -127,7 +127,9 @@ List new features before bug fixes.
...
+  `STRING_AGG`, `JSON_AGG`, `JSON_OBJECT_AGG`, `LIST_AGG`, or `ARRAY_AGG`
+  on data sets containing exactly one record.
```
Comment (asenac) [inline doc/user/content/release-notes.md:132]: This is missing something like: `... when that condition is known at optimization time`.

**Review by asenac — APPROVED** [review]
Very neat!

Just a comment about the release note.

Also, if this is going to be merged using a `merge` commit, I'd add a more descriptive commit message for this commit.

File: doc/user/content/release-notes.md (hunk 3 above)
Comment (mkysel) [reply doc/user/content/release-notes.md:132]: sure. Will do!

**Comment by philip-stoev** [comment]
For the record, this branch fixes a wrong result bug:
```
materialize=> SELECT  JSONB_AGG(a2 . f2) AS agg2 FROM ( SELECT LIST_AGG (f1) AS f1 FROM t1) AS a1 LEFT JOIN t2 AS a2 ON ( FALSE );
  agg2  
--------
 [null]
(1 row)
```

The prior commit was returning `[[null]]` , which did not match Postgres.

**Comment by philip-stoev** [comment]
This panic is also fixed:
```
2021-11-09T10:28:13.939195Z ERROR panic: Datum::unwrap_array called on List([Array(Array { elements: [Null], dims: [ArrayDimension { lower_bound: 1, length: 1 }] })])
thread: coordinator
location: src/expr/src/scalar/func.rs:5177:16
version: 0.9.12-dev (6df4866ed06653962c94acb37a0aed960516069b)
```

**Review by philip-stoev — APPROVED** [review]
I added the aggregate functions and heavy use of constant sources to the randomized grammar and no crashes or wrong results specific to this branch came out. On the contrary, there were wrong results and panics in main that are fixed.

**Pull request merged**